### PR TITLE
Fixes #17148: Allow searching hosts by hypervisor or guest

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

**Comment by ehelms** [comment]
@thomasmckay please review this

Note that this splits the original problem in half and attempts to only tackle the searching by hypervisor or guest types. The addition of being able to search for guests of a particular hypervisor adds additional complexity that will be handled in a separate PR and issue.

**Comment by ehelms** [comment]
A thought for reviewers on design, in addition to this, we could seed the application with a pre-created bookmark that searches for hypervisors and one that searches for guests to make it more discoverable.

File: db/seeds.d/108-subcription-bookmarks.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# This file should contain all the record creation needed to seed the database with its default values.
+# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
+#
+# !!! PLEASE KEEP THIS SCRIPT IDEMPOTENT !!!
+#
+
+Bookmark.without_auditing do
```
Comment (ehelms) [inline db/seeds.d/108-subcription-bookmarks.rb]: Added creation of two bookmarks here for the user that lists hyprvisors and list guests. Feel free to suggest better names for the bookmarks.

File: app/models/katello/host/subscription_facet.rb [hunk 2]
```diff
@@ -34,6 +35,19 @@ def import_database_attributes(consumer_params)
         end
       end
 
+      def update_consumer_type(consumer_params)
+        if consumer_params.try(:[], 'type').try(:[], 'label') == 'hypervisor'
+          self.consumer_type = 'hypervisor'
+        elsif consumer_params.try(:[], 'facts').try(:[], 'virt.is_guest') == "true"
```
Comment (thomasmckay) [inline app/models/katello/host/subscription_facet.rb]: I don't think we want to flag plain VMs as guests. The guests search should be for guests of hypervisors.

Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: So they are only guests if a hypervisor has checked in and reported them? That certainly complicates this.

Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: I think I would punt to this PR only reporting hypervisors, and a second PR focused on processing virt-who checkins to establish hypervisor and guest relationships.

Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: @jlsherrill thoughts since you were using this same fact for some logic

Comment (jlsherrill) [inline app/models/katello/host/subscription_facet.rb]: I think i'd disagree with you @thomasmckay.  A guest is a guest, it doesn't matter if we know about its hypervisor or not?

I've heard complaints from users about not knowing if a guest's hypervisor is registered or not due to virt-who miconfiguration which makes me think they would want to consider them guests.  Why wouldn't you?

File: app/models/katello/concerns/subscription_facet_host_extensions.rb [hunk 3]
```diff
@@ -18,6 +18,8 @@ module SubscriptionFacetHostExtensions
 
         has_many :activation_keys, :through => :subscription_facet
         has_one :subscription_status_object, :class_name => 'Katello::SubscriptionStatus', :foreign_key => 'host_id'
+        has_one :hypervisor, :through => :subscription_facet
```
Comment (jlsherrill) [inline app/models/katello/concerns/subscription_facet_host_extensions.rb]: did you mean to leave this here?

Comment (ehelms) [inline app/models/katello/concerns/subscription_facet_host_extensions.rb]: Nope!

File: app/models/katello/host/subscription_facet.rb (hunk 2 above)
Comment (thomasmckay) [inline app/models/katello/host/subscription_facet.rb]: I'd like to be able to search for "VMs with no hypervisor" and "VMs with a hypervisor". If you want to call all VMs guests, I guess that's alright but we use the terms hypervisor and guest in addition to physical and virtual. To me guest and virtual are different: A guest is a virtual with a known hypervisor.

Comment (jlsherrill) [inline app/models/katello/host/subscription_facet.rb]: I agree those are important queries to be available.  Eric's 2nd issue he is working on is to allow searching a guest by its hypervisor.  That would facilitate this search so you could do something like:

consumer_type = guest and set? hypervisor
or
consumer_type = guest and null? hypervisor

Comment (jlsherrill) [inline app/models/katello/host/subscription_facet.rb]: and i think those would make great bookmarks

Comment (thomasmckay) [inline app/models/katello/host/subscription_facet.rb]: Sounds fair, thanks

Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: If physical and virtual are important, and distinct from hypervisor and guest then lets agree to that and let me build that into this design. 

If the above is true, what would we call physical vs virtual vs hypervisor vs guest? I don't feel like a single field would encapsulate all those concerns. We could do the following fields (granted this adds more columns):

```
physical (boolean)
virtual (boolean)
consumer_type (string) values: hypervisor or guest
```

The `consumer_type` is very clearly a subscription facet, however, the physical and virtual get into the realm of the host properties directly. Today these values could be searched by facts with pre-made bookmarks to help users however that is not as obvious if we consider these attributes to be first class citizens of what a host is.

Comment (ohadlevy) [inline app/models/katello/host/subscription_facet.rb]: note that you can also use host.uuid to store its a guest, as that's with compute_resource_id later on maps to a compute resource. (if we get the uuid at this stage, we would be able to continue the mapping on the foreman side as well).

File: app/models/katello/host/subscription_facet.rb [hunk 4]
```diff
@@ -34,6 +35,19 @@ def import_database_attributes(consumer_params)
...
+          self.consumer_type = 'guest'
+        elsif !consumer_params.try(:[], 'guestIds').empty?
```
Comment (ohadlevy) [inline app/models/katello/host/subscription_facet.rb]: or even, just have a .guest? and a .hypervisor? methods

File: app/models/katello/host/subscription_facet.rb (hunk 2 above)
Comment (ohadlevy) [inline app/models/katello/host/subscription_facet.rb]: so you might prefill the uuid field to know its a vm, but set the consumer_type to guest only when there is a known mapping?

File: db/seeds.d/108-subcription-bookmarks.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  bookmarks = [
+    {:name => "list hypervisors", :query => 'consumer_type = hypervisor', :controller => "hosts"},
+    {:name => "list guests", :query => 'consumer_type = guest', :controller => "hosts"}
+  ]
+
+  bookmarks.each do |input|
+    next if Bookmark.where(:controller => input[:controller], :name => input[:name]).first
```
Comment (ohadlevy) [inline db/seeds.d/108-subcription-bookmarks.rb]: nitpick: can use .exists?

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    def report_errors
+      if @errors.any?
+        filename = "subscription_facet_upgrade-#{Time.now.to_i}.log"
+        path = "/var/log/foreman/#{filename}"
+        path = "/tmp/#{filename}" unless File.writable?(path)
+
+        file = File.open(path, 'w')
+        @errors.each { |error| file.write(error) }
```
Comment (ohadlevy) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: another nitpick, but I would just skip inserting \n and join the array with \n at this stage.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        file.close
+        $stderr.print "***********************************\n"
```
Comment (ohadlevy) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: nitpick: can use warn

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    @errors ||= []
+    User.current = User.anonymous_api_admin
+    puts _("Updating backend data for subscription facets")
+
+    #there may be some invalid hosts, if there are create a primary interface
+    ::Host.includes(:interfaces).find_each do |host|
+      if host.primary_interface.nil?
+        host.interfaces.create!(:primary => true)
```
Comment (ohadlevy) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: why is this here and not in a migration?

Comment (ohadlevy) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: imho this is totally unexpected and should be resolved in core.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      end
+    end
+
+    Katello::Host::SubscriptionFacet.find_each do |subscription_facet|
+      begin
+        candlepin_attrs = subscription_facet.candlepin_consumer.consumer_attributes
+        subscription_facet.import_database_attributes(candlepin_attrs)
+        subscription_facet.host = ::Host::Managed.find(subscription_facet.host_id)
```
Comment (ohadlevy) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: why do you need to find the host again? use
```ruby
subscription_facet.host_id = subscription_facet.host_id
```
or
```ruby
subscription_facet.host = subscription_facet.host
```

File: app/models/katello/host/subscription_facet.rb (hunk 4 above)
Comment (ohadlevy) [inline app/models/katello/host/subscription_facet.rb]: just wondering why this is not in a fact parser, similar to how we parse other facts.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake (hunk 8 above)
Comment (ares) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: How this happened in the first place? Primary interface should exist for every host now. Creating it this way might fail because of some interfaces validation added later.

Core migration that should have taken care of it
* https://github.com/theforeman/foreman/blob/develop/db/migrate/20140910153654_move_host_nics_to_interfaces.rb
* https://github.com/theforeman/foreman/blob/develop/db/migrate/20150508124600_copy_unmanaged_hosts_to_interfaces.rb

File: lib/katello/tasks/update_subscription_facet_backend_data.rake (hunk 9 above)
Comment (ShimShtein) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: Without looking into `#import_database_attributes`, I would say that the id is set, but the AR object is not retrieved. I am pretty sure that it's not necessary though...

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    puts _("Updating backend data for subscription facets")
+
+    #there may be some invalid hosts, if there are create a primary interface
+    ::Host.includes(:interfaces).find_each do |host|
+      if host.primary_interface.nil?
+        host.interfaces.create!(:primary => true)
+      end
+    end
+
+    Katello::Host::SubscriptionFacet.find_each do |subscription_facet|
```
Comment (ShimShtein) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: Do you have subscription facets without hosts?
IMHO you could rewrite it to:

```ruby
Katello::Host::SubscriptionFacet.includes(:host).find_each
```

It should set the `#host` automatically.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+      begin
+        candlepin_attrs = subscription_facet.candlepin_consumer.consumer_attributes
```
Comment (ShimShtein) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: Maybe it should be hidden inside `SubscriptionFacet`?

```ruby
class SubscriptionFacet
  def sync_from_candlepin
    attrs = candlepin_consumer.consumer_attributes
    import_database_attributes(attrs)
  end
end

File: app/models/katello/host/subscription_facet.rb (hunk 2 above)
Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: Where does the uuid field come from at this point? If I am client registering via sub-man, what is sending the VM uuid? (not to be confused with the subscription facet uuid coming from Candlepin)

Also, if we have the uuid to indicate its a VM, thats one thing. But are there existing ways for a user to search? e.g. `virtual = true` or `physical = true` ?

Comment (ohadlevy) [inline app/models/katello/host/subscription_facet.rb]: I'm uncertain, but if subman sends bios facts, those might contain the vm uuid.
regarding searching, if its puppet facts, then its facts.virtual = true/false.

File: db/seeds.d/108-subcription-bookmarks.rb (hunk 5 above)
Comment (ehelms) [inline db/seeds.d/108-subcription-bookmarks.rb]: Sure - to be fair, I just borrowed this seed from core

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  task :update_subscription_facet_backend_data => ["environment"] do
+    def error(message)
+      @errors << message
+      @errors << "\n"
+    end
+
+    def report_errors
+      if @errors.any?
+        filename = "subscription_facet_upgrade-#{Time.now.to_i}.log"
+        path = "/var/log/foreman/#{filename}"
```
Comment (ehelms) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: I'd note that this is simply a move of an existing file to a new location so it could be called in a shared manner. I'm hesitant to re-factor this code here since it is/has been used with the 3.0 upgrades and will be used again. A stand-alone update to this upgrade task to re-factor it might make more sense.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        candlepin_attrs = subscription_facet.candlepin_consumer.consumer_attributes
+        subscription_facet.import_database_attributes(candlepin_attrs)
+        subscription_facet.host = ::Host::Managed.find(subscription_facet.host_id)
+        subscription_facet.save!
+
+        host = subscription_facet.host
+        host.name = ::Katello::Host::SubscriptionFacet.sanitize_name(host.name)
+        host.save! if host.name_changed?
+
+        Katello::Host::SubscriptionFacet.update_facts(subscription_facet.host, candlepin_attrs[:facts])
```
Comment (ShimShtein) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: If we create a fact importer that accepts the whole `candlepin_attrs` hash as an input, we can avoid this update process altogether:
For each subscription facet: subscription_facet.host.import_facts(subscription_facet.candlepin_consumer.consumer_attributes)
The importer will take care for host name, and the import process will call the subscription facet to update its own fields from those facts.

IMHO this will be a bit straightforward solution.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake (hunk 8 above)
Comment (ehelms) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: @jlsherrill do you mind jumping in to comment on this artifact?

Comment (jlsherrill) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: Also a lot of these hacky workaround are due to the fact that hosts can be created or edited and put into an invalid state.  We probably should have just put 'save(:validate => false)' everywhere we needed to make sure a host got updated but so far we've tended to avoid that.  Not being able to count on the host object to valid has consistently been a very difficult issue.

File: app/models/katello/host/subscription_facet.rb (hunk 4 above)
Comment (jlsherrill) [inline app/models/katello/host/subscription_facet.rb]: We do utilize a fact parser here:  https://github.com/Katello/katello/blob/master/app/models/katello/rhsm_fact_parser.rb and likely could use that parser to do this functionality. 

So far we've only utilized the FactParser for attributes that the normal fact import process handles (such as operatingsystem, interfaces, etc..).  But no reason we couldn't use it for other things too.

File: app/models/katello/host/subscription_facet.rb (hunk 2 above)
Comment (jlsherrill) [inline app/models/katello/host/subscription_facet.rb]: ah after digging around i guess not

File: lib/katello/tasks/update_subscription_facet_backend_data.rake (hunk 8 above)
Comment (ares) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: We tried to change this on host importing where possible, but you're right, there're still ways to make persisted host invalid quite easily. Anyway, I think this should be removed and if it turns out that some host does not have primary interface (through exception on saving), please ping me so I can help debug. I'm afraid that keeping this could potentially mask new bugs.

Comment (ShimShtein) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: >  think this should be removed and if it turns out that some host does not have primary interface (through exception on saving), please ping me so I can help debug. I'm afraid that keeping this could potentially mask new bugs.

I would suggest leaving the code, but logging every parameter that could help. I am not sure what incentive would be enough for the user to actually send us those logs.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake (hunk 13 above)
Comment (ShimShtein) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: I am talking about `import_database_attributes` call a few lines earlier. Basically it would simplify the whole `find_each` block:

``` ruby
Katello::Host::SubscriptionFacet.find_each do |subscription_facet|
  begin
    candlepin_attrs = subscription_facet.candlepin_consumer.consumer_attributes
    subscription_facet.host.import_facts(candlepin_attrs)
  end
end
```

I was talking about https://github.com/theforeman/foreman/blob/develop/app/models/concerns/facets/base.rb#L47 that should be called by the framework automatically. I think `import_database_attributes` belongs there.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake (hunk 8 above)
Comment (ares) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: I don't think users reads log until there's an error

Comment (ohadlevy) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: @ares @ShimShtein how about a rake task (or UI) that can:
1. report all incorrectly defined hosts.
2. try to fix the problem (using facts?)
3. potentially force changing the host to unmanaged? (or some OK to keep state?)
4. try to help the user fix the problem?

File: app/models/katello/host/subscription_facet.rb (hunk 2 above)
Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: This conversation got deep into whether we include the virtual vs. physical nature of a host as part of this. That feels like it should be an automatic, separate change to all of this. And to keep this as focused on consumer type to allow users to find hypervisors and guests. I'd prefer to just open a separate issue for that ?

File: app/models/katello/host/subscription_facet.rb (hunk 4 above)
Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: I am not sure I got a strong feeling here one way or the other. This could always be re-factored to use this fact parser idea in the future.

**Comment by ehelms** [comment]
I'd prefer not to address all of the comments on the rake task I simply moved from one location to another in this PR.

**Comment by thomasmckay** [comment]
This does not work as-is. The call to update_consumer_type() during a "subscription-manager register" does not include the facts.

A katello:reimport did not update the consumer_type to guest either.

File: app/models/katello/host/subscription_facet.rb [hunk 14]
```diff
@@ -35,6 +36,19 @@ def import_database_attributes(consumer_params)
         end
       end
 
+      def update_consumer_type(consumer_params)
+        if consumer_params.try(:[], 'type').try(:[], 'label') == 'hypervisor'
+          self.consumer_type = 'hypervisor'
+        elsif host.facts.try(:[], 'virt.is_guest') == "true"
```
Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: @thomasmckay @jlsherrill  I switched this to pull from the `host.facts` as they appear to be get updated before this would be called and I guess should be the canonical source of information for this field.

File: lib/katello/tasks/update_subscription_facet_backend_data.rake (hunk 8 above)
Comment (ares) [inline lib/katello/tasks/update_subscription_facet_backend_data.rake]: I think the discussion diverged a bit here. What @jlsherrill said in second comment is true for any invalid host but the original concern was for hosts without the primary interface which I think is not easy to cause and generally should not happen. I'm not sure how often will this task be run, probably just once during the upgrade so I'm OK with keeping the hack there. But if it gets copied for each new task I think the code will be full of hacks no one understands :-)

**Comment by thomasmckay** [comment]
From IRC discussion, I _think_ this is what we arrived at:

1. use existing logic only to set a new searchable column "is_hypervisor = true/false"
2. add new search column "hypervisor = $name", and "hypervisor = null" to find VMs that belong to a hypervisor or not

**Comment by jlsherrill** [comment]
I'm not a huge fan of 'is_hypervisor' but don't really have any alternative if we are going with true/false for hypervisor.  One clarification to search for null items you actually use:  `null? hypervisor`

**Comment by ehelms** [comment]
A few options:

#1 
```
is_hypervisor = true/false
hypervisor = name/null
```

#2 
```
hypervisor = true/false
hypervisor_hostname = name/null
```

#3 
```
hypervisor = true/false
hypervisor_name = name/null
```

**Comment by thomasmckay** [comment]
#4
hypervisor = true/false
guests_of = $hostname/null

#5
hypervisor = true/false
guests_of_hypervisor = $hostname/null

**Comment by jlsherrill** [comment]
I kinda like 3 or 5

**Comment by thomasmckay** [comment]
My vote would be for #5

**Comment by ares** [comment]
Sorry for introducing #6 

virtualization_host = true/false
virtualization_guest = true/false
hypervisor = hostname/null

If you don't like it, I'd prefer #3

**Comment by ehelms** [comment]
With `hypervisor` as a true/false field on the subscription_facet model, the challenge becomes how to store the hypervisor/guest relationship. The natural inclination would be to do:

```
has_one :hypervisor
```

However, that won't work. So I need some thoughts on how to store this relationship, some options:

One each guest:
```
has_one :host_hypervisor
```
```
has_one :hypervisor_host
```

Or, on the hypervisors:
```
has_many :guests
```

**Comment by ShimShtein** [comment]
@ehelms I would suggest moving hypervisor-guest relationship outside of subscription facet, and putting it directly on host (as extension) as a many-to-many relation table (has one hypervisor, has many guests).
I know that most of the usage is for subscriptions, but it sounds to me like something a bit more generic. IMHO subscription facet is a place for properties tightly related to the subscription, like id, type of the subscription e.t.c. and less for managing relations between hosts.

What do you think?

**Comment by ehelms** [comment]
@ShimShtein The thing is, without subscriptions, there is not a strong need for this hypervisor/guest relationship searching that we are putting into place. The relationship is reported by virt-who and lives in Candlepin which tightly couples it to the subscription aspect.

**Comment by ShimShtein** [comment]
In this case I would suggest maybe a different naming inside subscription facet?

``` ruby
has_one :subscribed_hypervisor

has_many :subscribed_guests
```

It represents the fact that this is not a general hosts relationship, but relationship between subscriptions. 
BTW, maybe it makes sense even to connect subscription_facets and not hosts. So effectively, you would manage subscriptions relationships, and not host's. You will still be able to get to the host object (facet has `.host`), but the main player would be the subscription.

**Comment by jlsherrill** [comment]
i'd rather avoid using 'subscribed' in the relationship name to not confuse it with subscriptions.  Connecting sub-facets is an interesting idea, i'm not sure that it gains us a lot though.

i'm fine with  'hypervisor_host' for the relationship name

**Comment by ShimShtein** [comment]
That's exactly my concern - if the relationship is about hosts it shouldn't be in the facet, but if the relationship is about how the host reports its subscription type (hypervisor or guest) then we should leave it in the facet and name it properly.

**Comment by ehelms** [comment]
@thomasmckay @jlsherrill This is now ready for re-review and includes the following functionality:

 1) ability to find hypervisors via searching "hypervisor = true"
 2) ability to find guests of a particular hypervisor via searching "hypervisor_host = HOSTNAME" (e.g. "hypervisor_host = example.org"

This should be handling properly updates coming from either the hypervisor or the guest and no matter what order they are presented to Katello in.

**Comment by ehelms** [comment]
I've updated this PR to incorporate changes from @thomasmckay -- you could ensure this works as you expect. @jlsherrill if you wouldn't mind doing the final review of mine + @thomasmckay changes.

**Comment by thomasmckay** [comment]
ack! awesome! want!

File: app/lib/actions/katello/host/update.rb [hunk 15]
```diff
@@ -12,11 +12,11 @@ def plan(host, consumer_params = nil)
...
 
             if host.subscription_facet
-              if consumer_params
-                host.subscription_facet.update_from_consumer_attributes(consumer_params)
-              else
+              unless consumer_params
                 consumer_params = host.subscription_facet.consumer_attributes
               end
+
+              host.subscription_facet.update_from_consumer_attributes(consumer_params)
```
Comment (jlsherrill) [inline app/lib/actions/katello/host/update.rb]: whats the purpose of this change?  Seems like its needlessly updating the subscription_facet here.

File: app/models/katello/concerns/subscription_facet_host_extensions.rb [hunk 16]
```diff
@@ -16,6 +16,8 @@ module SubscriptionFacetHostExtensions
 
         has_many :activation_keys, :through => :subscription_facet
         has_one :subscription_status_object, :class_name => 'Katello::SubscriptionStatus', :foreign_key => 'host_id'
+        has_one :hypervisor_host, :through => :subscription_facet
```
Comment (jlsherrill) [inline app/models/katello/concerns/subscription_facet_host_extensions.rb]: i don't think this is needed anymore since you're using an ext_method for searching?

Comment (ehelms) [inline app/models/katello/concerns/subscription_facet_host_extensions.rb]: This was still throwing errors without it when searching from the UI

File: app/lib/actions/katello/host/update.rb (hunk 15 above)
Comment (ehelms) [inline app/lib/actions/katello/host/update.rb]: This was based on feedback from @thomasmckay here (https://github.com/ehelms/katello/pull/16/files/c4e54e9faef0fe6de572efba3aa99431fdbd7f64#diff-0814801452c6a2779e2bf22963f6bd58) related to his testing trying to do updates via host API.

Comment (thomasmckay) [inline app/lib/actions/katello/host/update.rb]: Either the consumer_params are passed in, or they are fetched (some attributes are stripped out in controller). This may result in extra updates but it will guarantee everything is in sync.

**Comment by ehelms** [comment]
@thomasmckay this went through a refactoring session with @jlsherrill to uncover issues he saw with virt-who/sub-man registration workflows. Could you please re-test?

File: app/models/katello/host/subscription_facet.rb [hunk 17]
```diff
@@ -35,6 +39,35 @@ def import_database_attributes(consumer_params)
...
+        elsif !candlepin_consumer.virtual_guests.empty?
+          # Check by calling out to Candlepin last for efficiency
+          self.hypervisor = true
+        end
+      end
+
+      def update_guests(consumer_params)
+        if self.hypervisor
+          if !consumer_params.try(:[], 'guestIds').empty?
+            guest_ids = consumer_params['guestIds'].map{ |i| i['guestId'].downcase}
```
Comment (thomasmckay) [inline app/models/katello/host/subscription_facet.rb]: consumer_params['guestIds'] is just an array of strings and i['guestId'] is nil

Comment (thomasmckay) [inline app/models/katello/host/subscription_facet.rb]: Why downcase here? These are IDs not hostnames.

Comment (thomasmckay) [inline app/models/katello/host/subscription_facet.rb]: I'm fine w/ downcase as it seems to work.

The guestIds come in through here https://github.com/Katello/katello/blob/master/app/controllers/katello/api/v2/host_subscriptions_controller.rb#L11

Comment (ehelms) [inline app/models/katello/host/subscription_facet.rb]: So this is a case where the Katello API is not mimicking the RHSM API in terms of the data format it expects? Or can you point me to an example where `virt-who` provided the `guestIds` field as an array of strings?

Comment (thomasmckay) [inline app/models/katello/host/subscription_facet.rb]: Try an older version of virt-who. Previously the json sent up was literally just: $hypervisor: [$id1, $id2, $id3]

Since this is our API I don't think we should change the format of a parameter. Perhaps accept both new and old and convert at the controller?

**Comment by ehelms** [comment]
@thomasmckay @jlsherrill Sorry this took so long but it has been updated to account for the array style of string UUIDs case.

**Comment by thomasmckay** [comment]
APJ

**Comment by jlsherrill** [comment]
APJ (small rubocop issue)
